libs/basics/src/result.ts

```diff
@@ -215,3 +232,87 @@ export type Result<T, E> = Ok<T> | Err<E>;
 export function isResult(value: unknown): value is Result<unknown, unknown> {
   return value instanceof Ok || value instanceof Err;
 }
+
+class ResultBlockError<E> {
+  constructor(private readonly error: E) {}
+
+  err(): E {
+    return this.error;
+  }
+}
+
+/**
+ * Provides an early-return mechanism for functions that return `Result`s,
+ * similar to Rust's `?` operator. This is useful for avoiding `if` statements
+ * that check for errors and return early.
+ *
+ * Note that errors thrown are not caught by this function, so you should use
+ * standard try/catch blocks to catch thrown errors.
+ *
+ * @example
+ *
+ * ```ts
+ * function doSomething(): Result<string, Error> {
+ *   return resultBlock((ret) => {
+ *     const value = doSomethingThatMightFail().or(ret);
+ *     const value2 = doSomethingElseThatMightFail(value).or(ret);
+ *     return anotherThingThatMightFail(value2);
+ *   });
+ * }
+ * ```
+ */
+export function resultBlock<T, E>(
+  fn: (ret: (error: E) => never) => T | Result<T, E>
+): Result<T, E> {
+  function ret(error: E): never {
+    throw new ResultBlockError(error);
+  }
+
+  try {
+    const returnValue = fn(ret);
+    return isResult(returnValue) ? returnValue : ok(returnValue);
+  } catch (error) {
+    if (error instanceof ResultBlockError) {
+      return err(error.err());
+    }
+    throw error;
+  }
+}
+
+/**
+ * Provides an early-return mechanism for async functions that return `Result`s,
+ * similar to Rust's `?` operator. This is useful for avoiding `if` statements
+ * that check for errors and return early.
+ *
+ * Note that errors thrown are not caught by this function, so you should use
+ * standard try/catch blocks to catch thrown errors.
+ *
+ * @example
+ *
+ * ```ts
+ * function doSomething(): Promise<Result<string, Error>> {
+ *   return asyncResultBlock(async (ret) => {
+ *     const value = (await doSomethingThatMightFail()).or(ret);
+ *     const value2 = (await doSomethingElseThatMightFail(value)).or(ret);
+ *     return anotherThingThatMightFail(value2);
+ *   });
+ * }
+ * ```
+ */
+export async function asyncResultBlock<T, E>(
```

Comment:
Is it possible to write one version of `resultBlock` that supports sync and async based on the type of the callback given?

Reply:
I tried this and could not get it to work. It's straightforward enough to build something that runs correctly, but making it so TypeScript can properly type check it seems hard. No matter what I tried it would always end up inferring the type parameters as `unknown`.